### ISSUE TITLE
nn.Embedding padding_idx with custom initialization does not work as expected

### ISSUE THREAD

**Issue by alexeib** [pytorch/pytorch#6411]
When initializing nn.Embedding with a padding_idx but then overwriting weights with some custom initialization, padding_idx no longer has any effect since the weights at that position are overridden. One would intuitively expect padding_idx to be handled on a higher level and always produce 0 embeddings for padding elements, rather than only during the weight initialization.

This seems dangerous and counterintuitive. Can we make padding_idx work properly, or at least document this in a very visible way?

See https://github.com/pytorch/pytorch/issues/1812 where somebody thought it was a bug

**Comment by SsnL** [comment]
We can't require `padding_idx` embedding to always be 0 because 
1. The weight matrix is just a tensor, which is the same as any other tensor. In fact, one can use any tensor with proper shape as the weight matrix (both in module form and functional form of embedding). So it will be difficult and doesn't make sense to impose such a constraint
2. Sometimes people may actually want to set `padding_idx` to some specific embedding.

I'll add a note about this in our document.

**Comment by myleott** [comment]
I think the point was that the current documentation for padding_idx says: "padding_idx (int, optional) – If given, pads the output with zeros whenever it encounters the index."

But that isn't quite true. It just zeros the weights for padding_idx in the constructor. If one re-initializes the weights (as is often done in practice, e.g., in the PyTorch word_language_model example [1]), then padding_idx is effectively ignored.

Updating the documentation will be helpful.

[1] https://github.com/pytorch/examples/blob/master/word_language_model/model.py#L41-L43

**Comment by apaszke** [comment]
Or maybe we should change `padding_idx` to mean that we will zero the gradient for that entry"?

**Comment by SsnL** [comment]
closed via #6430

**Comment by apaszke** [comment]
@SsnL thanks!
